### PR TITLE
Adding-Modification-Functionality-For-Checkout-Api

### PR DESCRIPTION
**Pull request by eraserhead0705**
What problem you’re solving:
The Adyen Python Library lacks support for the full Checkout API actions.
Capture
Refund
Reversals
Cancel/Technical Cancel

Our Team at Tier needs this functionality implemented in the Python library in order to fully integrate with our backend system.

Your approach to fixing the problem:
Implemented the above functionality by building on the current service/client in the code base. I added a new parameter to send a path parameter to the Adyen Api.

Any tests you wrote:
Each new action has two tests written, one success and one that mocks an error.
Additionally a test that checks for race condition.

### CODE REVIEW

**Comment by eraserhead0705** [comment]
@acampos1916 Hey! This is a reopened PR per your instructions. I made the requested changes. Thanks again!

File: test/CheckoutTest.py [hunk 1]
```diff
@@ -236,6 +236,201 @@ def test_payments_result_error_mocked(self):
...
+            "reference": "YourCancelReference",
+        }
+        psp_reference = "Payment123"
+        self.adyen.client = self.test.create_client_from_file(200, requests,
+                                                              "test/mocks/"
+                                                              "checkout/"
+                                                              "paymentscancel-"
+                                                              "withoutreference-succes.json")
+
+        result = self.adyen.checkout.payments_cancels_without_reference(request=requests, path_param=psp_reference)
```
Comment (Aleffio) [inline test/CheckoutTest.py:314]:
```suggestion
        result = self.adyen.checkout.payments_refunds(request=requests, path_param=psp_reference)
```

**Comment by AlexandrosMor** [comment]
Hello @eraserhead0705,
Thank you for the pull request. We see that there are old commits which is not good for the commit history. Would you mind to take care of it ? 

Thank you in advance, 
Alexandros
Adyen

File: Adyen/client.py [hunk 2]
```diff
@@ -171,6 +172,14 @@ def _determine_checkout_url(self, platform, action):
             action = "payments/details"
         if action == "paymentsResult":
             action = "payments/result"
+        if action == "paymentsCancelsWithoutReference":
+            action = "payments/cancels"
```
Comment (Aleffio) [inline Adyen/client.py:176]: are you referring to this endpoint? https://docs.adyen.com/api-explorer/#/CheckoutService/v67/post/cancels
it should be just `/cancels`? without the `payments` prefix. Maybe we can call the action just `cancels` if this is the case.

File: test/CheckoutTest.py [hunk 3]
```diff
@@ -236,6 +236,201 @@ def test_payments_result_error_mocked(self):
...
+            "reference": "YourCancelReference",
+        }
+        reference_id = "Payment123"
+        self.adyen.client = self.test.create_client_from_file(200, requests,
+                                                              "test/mocks/"
+                                                              "checkout/"
+                                                              "paymentsresult-error-invalid-"
+                                                              "data-payload-422.json")
+
+        result = self.adyen.checkout.payments_cancels_without_reference(request=requests, path_param=reference_id)
```
Comment (Aleffio) [inline test/CheckoutTest.py:332]:
```suggestion
        result = self.adyen.checkout.payments_refunds(request=requests, path_param=reference_id)
```

File: test/CheckoutTest.py [hunk 4]
```diff
@@ -236,6 +236,201 @@ def test_payments_result_error_mocked(self):
...
+        self.assertEqual(422, result.message['status'])
+        self.assertEqual("14_018", result.message['errorCode'])
+        self.assertEqual("Invalid payload provided", result.message['message'])
+        self.assertEqual("validation", result.message['errorType'])
+
+    def test_payments_refunds_raises_vaulue_error(self):
```
Comment (Aleffio) [inline test/CheckoutTest.py:338]:
```suggestion
    def test_payments_refunds_raises_value_error(self):
```

File: test/CheckoutTest.py [hunk 5]
```diff
@@ -236,6 +236,201 @@ def test_payments_result_error_mocked(self):
...
+            "merchantAccount": "YOUR_MERCHANT_ACCOUNT",
+            "reference": "YourCancelReference",
+        }
+        self.adyen.client = self.test.create_client_from_file(200, requests,
+                                                              "test/mocks/"
+                                                              "checkout/"
+                                                              "paymentscancel-"
+                                                              "withoutreference-succes.json")
+        with self.assertRaises(ValueError) as exc:
+            self.adyen.checkout.payments_cancels_with_reference(request=requests, path_param="")
```
Comment (Aleffio) [inline test/CheckoutTest.py:350]:
```suggestion
            self.adyen.checkout.payments_refunds(request=requests, path_param="")
```

File: Adyen/services.py [hunk 6]
```diff
@@ -322,6 +329,42 @@ def payment_result(self, request=None, **kwargs):
         action = "paymentsResult"
         return self.client.call_checkout_api(request, action, **kwargs)
 
+    def payment_captures(self, path_param, request=None, idempotency_key=None, **kwargs):
+        if path_param == "":
+            raise ValueError(
+                'must contain a pspReference in the path_param, path_param cannot be empty'
+            )
+        action = "paymentsCapture"
```
Comment (Aleffio) [inline Adyen/services.py:337]: can you please add this action in the `client.py` code? I think it's not there yet

File: Adyen/services.py [hunk 7]
```diff
@@ -322,6 +329,42 @@ def payment_result(self, request=None, **kwargs):
         action = "paymentsResult"
         return self.client.call_checkout_api(request, action, **kwargs)
 
+    def payment_captures(self, path_param, request=None, idempotency_key=None, **kwargs):
```
Comment (Aleffio) [inline Adyen/services.py:332]:
```suggestion
    def payments_captures(self, path_param, request=None, idempotency_key=None, **kwargs):
```

**Comment by acampos1916** [comment]
Thanks @eraserhead0705 let us know it goes, we can check the PR again as soon as these comments are addressed.

File: Adyen/services.py [hunk 8]
```diff
@@ -322,6 +329,42 @@ def payment_result(self, request=None, **kwargs):
...
+        action = "paymentsCancelsWithReference"
+        return self.client.call_checkout_api(request, action, path_param, idempotency_key, **kwargs)
+
+    def payments_reversals(self, path_param, request=None, idempotency_key=None, **kwargs):
+        if path_param == "":
+            raise ValueError(
+                'must contain a pspReference in the path_param, path_param cannot be empty'
+            )
+        action = "paymentsReversals"
+        return self.client.call_checkout_api(request, action, path_param, idempotency_key, **kwargs)
```
Comment (whyscream) [inline Adyen/services.py:358]: This does not work: `call_checkout_api` expects `idempotency_key` and `path_param` in a different order. 

Same comment probably also applies to all other uses of `call_checkout_api` in this PR.

Comment (eraserhead0705) [reply Adyen/services.py:358]: I see :) Thanks for the comment!

**Review by cyattilakiss — APPROVED** [review]

**Comment by yesthesoup** [comment]
hi all, my team would like to use this as well - what is blocking this PR? maybe we could help contribute

**Comment by acampos1916** [comment]
Hello @yesthesoup, your help would be greatly appreciated. I see that @eraserhead0705 hasn’t been able to work on the suggestions from above, are you able to commit to this branch?

**Comment by eraserhead0705** [comment]
Hey @acampos1916 I have the fixes on my local branch. I will try and get them pushed this week.

**Comment by acampos1916** [comment]
Thanks @eraserhead0705, if it’s easier to open a new one feel free to do so.

**Comment by eraserhead0705** [comment]
> Thanks @eraserhead0705, if it’s easier to open a new one feel free to do so.

Gonna open a new one then, in the end I think its cleaner.

**Comment by eraserhead0705** [comment]
@acampos1916 I opened a New PR, https://github.com/Adyen/adyen-python-api-library/pull/158 for these changes.

**Comment by eraserhead0705** [comment]
@acampos1916 Can I close this PR?

**Comment by acampos1916** [comment]
Thanks @eraserhead0705, I can close it

**Pull request closed without merging**